### ISSUE TITLE
Specify the kind of a .tres with a .kind.tres "double extension" filenaming scheme

### ISSUE THREAD

**Issue by duchainer** [godotengine/godot-proposals#4668]
### Describe the project you are working on

None related to this, as it was mostly discussed in 2022-06-14 's PR review meeting when reviewing  https://github.com/godotengine/godot/pull/61539.

### Describe the problem or limitation you are having in your project

We do want:
 - a standard naming for subtypes of resource files (there after called kind)
 - to keep the ResourceLoader simple enough without having to deal with multiple extension for a single type
 - a way to tell at a glance what kind is a file from the file explorers
 - a faster way to select which icon to display next to a .tres file in the FileSystem as @YuriSizov mentioned
   - It would allow file managers to pick an icon depending on that double extension like for javascript files
 
 Also @reduz did propose that we have new extensions for visual code and shaders: https://github.com/godotengine/godot-proposals/issues/3546
 So that would allow for .vsh.tres (visual shader) and .vs.tres (visual script) too.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

(Open to bikesheding it in the comments:)
Using a .`filename.kind.tres` filename format does allow for :
- .mat.tres
- .vsh.tres (visual shader)
- .mesh.tres
- .env.tres
- .font.tres
- .style.tres
- .theme.tres 
- and the rest would just fallback to plain .tres
If you think of any others, please comment below and I'll add them to the list

So it is an really extensible way of documenting the inner kind of a Resource.
It does not even have to restrict the kind naming, so a project could use :
- .action.tres 
if they had a ActionData Resource sub class, like GDQuest Turn based example: https://github.com/GDQuest/godot-2d-jrpg-combat/blob/b01bce7712cd67f3f8c66abcb786a9792872b4ae/godot/CombatSystem/Battler/Actions/ActionData.gd

That would also allow to assert that the file kind is what we want, so you would be able to check if that resource is a material or a texture resource from the filename itself first.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

So it would be adding the .kind.tres instead of only .tres when saving a resource of that kind (like a visual shader, or a dynamic font, etc)

It would store a key-value pair of some sort that match the type(s) of the created Resource with the kind name for the file.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Currently we can name files with .kind.tres, so it is rather to have a standard that is applied on Resource when saving to disk.

### Is there a reason why this should be core and not an add-on in the asset library?

I don't think that we have access to the Resource creation dialog from an EditorPlugin, and it is somewhat of a standardizing question too.

**Comment by Zireael07** [comment]
What about converting existing .tres files?

**Comment by duchainer** [comment]
I suppose that would be rather re-saving those, either:
 - renaming them (which I don't think the Save dialog does allow), or;
 - duplicating them.
 But they can still be used the same, so it can be done as progressively as needed I'd think.

**Comment by lyuma** [comment]
Just to be clear, the proposal is simply a change to the "save" dialog, right? (As well as some mapping from Resource class to default extension) It means that the save-as dialog will pre-fill the resource type in the box, but nothing stops you from using a plain `.tres`

> What about converting existing .tres files?

If I understand, this is only a convention, not a hard requirement, so I don't think converting existing .tres files should be required. If the user wishes to rename them manually they can. (Also, if the `FileSystemDock::_rename_operation_confirm` function eventually gets exposed to `EditorFileSystem` for scripts to use, it might be possible to automate the conversion as an addon, if desired.)

**Comment by duchainer** [comment]
@lyuma Yeah, it is really more supporting a convention in the Save dialog, not at all adding requirement on the engine.
Like how we currently have snake_cased names by default for new files but there is no hard requirement for snake_casing in the engine.
